### PR TITLE
[#1021, #2978] Automatically calculated spell preparation limits

### LINKED ISSUE #1021: Add per-class spell preparation tracking
Originally in GitLab by @Kandashi

Adding an X/Y for attuned items and prepared spells would be a nice QoL feature. No hard limits, just a visual reminder for the player. Possibly with a Special Trait for manual increasing of these limits (artificers etc)

Priority low

some additional info: 
```
honeybadger (he/him)Today at 3:38 PM
since the class item already has spell progression, you wouldnt even need to hardcode it
full = level+mod
half = level/2 (round down) + mod
```

maybe an addon to #1012

### LINKED ISSUE #2978: Add display of number of prepared spells back to spellbook
The old character sheet displayed the number of prepared spells in the header. This number is no longer shown in the new character sheet.
This was useful for wizards and clerics, who can know any number of spells, but only have access to a limited number of them at one time, dictated by level and ability modifier.
As a bonus feature, which would require significantly more work, it would be great if this count was separate for each class. That is, a multiclassed cleric/wizard would have two prepared skill counts. That means that each spell in the list would need to be qualified as wizard or cleric. This is easy for spells that only one class has access to (magic missile and cure wounds, for example), but would require the user to select a class for spells that both classes have access to, like detect magic, or wizard spells that certain cleric domains have access to.

### PR DESCRIPTION
**Pull request by absymylyar**
- Move the spellcasting input to a common template for class and subclass
- Add the "needSpellPreparation" field to spellcasting data
- Calculate and display the spell preparation limit

Closes #1021 
Closes #2978

### CODE REVIEW

File: module/documents/item.mjs [hunk 1]
```diff
@@ -392,6 +392,17 @@ export default class Item5e extends SystemDocumentMixin(Item) {
...
+      && CONFIG.DND5E.spellcastingTypes.leveled.progression[finalSC.progression]) {
+      finalSC.spellPreparationLimit = Math.max(
+        1,
+        this.parent.system.abilities[finalSC.ability].mod
+        + Math.ceil(
+          this.system.levels / (CONFIG.DND5E.spellcastingTypes.leveled.progression[finalSC.progression].divisor || 1)
+        )
+      );
+    }
+
```
Comment (arbron) [inline module/documents/item.mjs:405]: This preparation would probably be better handled within the `prepareFinalData` method on class & subclass items. We could add a static method to `SpellcastingField` to perform the calculation and then call that from the class & subclass data models:

Added to `ClassData` and `SubclassData` after the "Data Migrations" section:
```javascript
  /* -------------------------------------------- */
  /*  Data Preparation                            */
  /* -------------------------------------------- */

  /** @inheritDoc */
  prepareFinalData() {
    SpellcastingField.calculatePreparationLimit(this.spellcasting, this.parent.getRollData({ deterministic: true });
  }
```

Added to `SpellcastingField`:
```javascript
  /**
   * Calculate total preparation limit.
   * @param {object} spellcasting  Spellcasting data for this object.
   * @param {object} rollData      Roll data for the actor being prepared.
   */
  static calculatePreparationLimit(spellcasting, rollData) {
    // Calculate preparation limit here using `simplifyBonus` utility
  }
```

```suggestion
```

File: packs/_source/classes/cleric.json [hunk 2]
```diff
@@ -389,7 +389,8 @@
     ],
     "spellcasting": {
       "progression": "full",
-      "ability": "wis"
+      "ability": "wis",
+      "needToPrepare": true
```
Comment (arbron) [inline packs/_source/classes/cleric.json:393]:
```suggestion
      "preparationFormula": "@abilities.wis.mod + @classes.cleric.levels"
```

File: packs/_source/classes/druid.json [hunk 3]
```diff
@@ -404,7 +404,8 @@
     ],
     "spellcasting": {
       "progression": "full",
-      "ability": "wis"
+      "ability": "wis",
+      "needToPrepare": true
```
Comment (arbron) [inline packs/_source/classes/druid.json:408]:
```suggestion
      "preparationFormula": "@abilities.wis.mod + @classes.druid.levels"
```

File: packs/_source/classes/paladin.json [hunk 4]
```diff
@@ -492,7 +492,8 @@
     ],
     "spellcasting": {
       "progression": "half",
-      "ability": "cha"
+      "ability": "cha",
+      "needToPrepare": true
```
Comment (arbron) [inline packs/_source/classes/paladin.json:496]:
```suggestion
      "preparationFormula": "@abilities.cha.mod + floor(@classes.paladin.levels / 2)"
```

File: packs/_source/classes/wizard.json [hunk 5]
```diff
@@ -348,7 +348,8 @@
     ],
     "spellcasting": {
       "progression": "full",
-      "ability": "int"
+      "ability": "int",
+      "needToPrepare": true
```
Comment (arbron) [inline packs/_source/classes/wizard.json:352]:
```suggestion
      "preparationFormula": "@abilities.int.mod + @classes.wizard.levels"
```

File: templates/items/parts/item-spellcasting.hbs [hunk 6]
```diff
@@ -15,3 +15,10 @@
         </select>
     </div>
 </div>
+
+<div class="form-group">
+    <label>{{localize "DND5E.SpellneedToPrepare"}}</label>
+    <div class="form-fields">
+        <input type="checkbox" name="system.spellcasting.needToPrepare" {{ checked system.spellcasting.needToPrepare }}>
```
Comment (arbron) [inline templates/items/parts/item-spellcasting.hbs:22]:
```suggestion
        <input type="string" name="system.spellcasting.preparationFormula" value="{{ system.spellcasting.preparationFormula }}">
```

File: templates/items/parts/item-spellcasting.hbs [hunk 7]
```diff
@@ -15,3 +15,10 @@
         </select>
     </div>
 </div>
+
+<div class="form-group">
+    <label>{{localize "DND5E.SpellneedToPrepare"}}</label>
```
Comment (arbron) [inline templates/items/parts/item-spellcasting.hbs:20]:
```suggestion
    <label>{{ localize "DND5E.SpellPreparation.Formula" }}</label>
```

File: lang/en.json [hunk 8]
```diff
@@ -1462,6 +1462,7 @@
 "DND5E.SpeedConditions": "Speed Conditions",
 "DND5E.SpeedSpecial": "Special Movement",
 "DND5E.SpellAbility": "Spellcasting Ability",
+"DND5E.SpellneedToPrepare": "Need to Prepare Spells",
```
Comment (arbron) [inline lang/en.json:1465]:
```suggestion
```

**Review by arbron — CHANGES_REQUESTED** [review]

**Comment by arbron** [comment]
Could you also comment on https://github.com/foundryvtt/dnd5e/issues/1021 & https://github.com/foundryvtt/dnd5e/issues/2978 so I can assign them to you?

**Comment by absymylyar** [comment]
Sorry, for late reply, I was very busy all those last weeks.
I take all dev in charge back

**Comment by arbron** [comment]
Modification of this merged: https://github.com/foundryvtt/dnd5e/pull/3811

**Pull request closed without merging**